{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d34ed3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading luad v3.1.1..                    \r"
     ]
    }
   ],
   "source": [
    "import cptac\n",
    "cptac.download(dataset=\"Luad\")\n",
    "lung = cptac.Luad()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abf63378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(211, 10699)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data = lung.get_proteomics() # get the proteomics data\n",
    "protein_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "adapted-union",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages here\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "promotional-ivory",
   "metadata": {},
   "source": [
    "### Short Pandas Review\n",
    "We can access the rows/column names of a data frame in two ways:\n",
    "\n",
    "1. We can get the row and column names together in a list with the `axes` attribute.\n",
    "2. We can get the row or column names individually in the `index` (row) or `columns` (column) attribute.\n",
    "\n",
    "\n",
    "**Exercise.** Print the `axes`, `index`, and `columns` of the `protein_data` object. Then, determine the dimensions of the `protein_data`, stored in the `shape` attribute. Finally, answer the following questions:\n",
    "\n",
    "- How many rows and columns are there in the data frame?\n",
    "- What do the rows and columns represent?\n",
    "- How many patients are there?\n",
    "- What about proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "expired-reach",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['C3L-00001', 'C3L-00009', 'C3L-00080', 'C3L-00083', 'C3L-00093',\n",
       "       'C3L-00094', 'C3L-00095', 'C3L-00140', 'C3L-00144', 'C3L-00263',\n",
       "       ...\n",
       "       'C3N-02423.N', 'C3N-02424.N', 'C3N-02433.N', 'C3N-02529.N',\n",
       "       'C3N-02572.N', 'C3N-02582.N', 'C3N-02586.N', 'C3N-02587.N',\n",
       "       'C3N-02588.N', 'C3N-02729.N'],\n",
       "      dtype='object', name='Patient_ID', length=211)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Explore the data frame here\n",
    "#protein_data.shape\n",
    "# 197 rows and 8067 columns\n",
    "\n",
    "#protein_data.index # The rows are the patient's ID\n",
    "#print(protein_data.columns[0])\n",
    "protein_columns = protein_data.columns\n",
    "new_protein_columns = list()\n",
    "for i in range(len(protein_columns)):\n",
    "    new_protein_columns.append(protein_columns[i][0])\n",
    "\n",
    "#print(new_protein_columns)\n",
    "\n",
    "#protein_patient_ID = list()\n",
    "#for i in range(len(protein_columns)):\n",
    "#    protein_patient_ID.append(protein_columns[i][1])\n",
    "\n",
    "#print(protein_patient_ID)\n",
    "\n",
    "protein_data.index\n",
    "    \n",
    "#print(protein_data.columns[1])\n",
    "# The columns are the name of the protein\n",
    "\n",
    "# There are 197 patients and 8067 proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add2d694",
   "metadata": {},
   "source": [
    "As for indexing, recall that `loc[]` accesses values by name, and `iloc[]` accesses values by index. `iloc[]` won't be too necessary for this data.\n",
    "\n",
    "**Exercise.** Access the following:\n",
    "1. Print the relative fold change of patient 01CO014 in the `CTNNB1` (beta-catenin) protein. Then, interpret this value -- does this patient have higher or lower protein levels than the reference? How much different is it? (Hint: use `loc[]`. Be sure to specify both the rows and the columns!)\n",
    "2. Access all values in the `COL4A1` column. Because this is a column, there's many ways to do so -- can you think of at least two?\n",
    "3. Finally, use boolean indexing to access the patient IDs with high (â‰¥1) `KRAS` expression:\n",
    "    a. Think about if KRAS expression information will be contained in a row or a column!\n",
    "    b. Create a boolean mask with the row/column that contains the expression information of KRAS.\n",
    "    c. Apply the mask to the row/column names that contains the patient IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eda8c90d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Access cell of patient 01CO014, for the gene CTNNB1\n",
    "part1 = protein_data.loc[\"01CO014\", \"CTNNB1\"]\n",
    "assert(part1 == -0.0431)\n",
    "\n",
    "# 2. Access all the values in the COL4A1 column\n",
    "part2 = protein_data[\"COL4A1\"]\n",
    "assert(len(part2) == 197)\n",
    "\n",
    "# 3. Get the patients with high KRAS expression\n",
    "KRAS_mask = protein_data[\"KRAS\"] > 1\n",
    "part3 = protein_data.index[KRAS_mask]\n",
    "assert(part3 == \"05CO028.N\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forbidden-advertiser",
   "metadata": {},
   "source": [
    "### Additional data in cptac\n",
    "You can also access other data using the `cptac` python package for the same patients, using the corresponding `get` function.\n",
    "\n",
    "**Exercise**. Access the RNA (`transcriptomics`) and `clinical` data from the colon object. The function name is very similar in syntax to the `get_proteomics()` function from before!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "13e51b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_data = lung.get_transcriptomics()\n",
    "rna_data.shape\n",
    "clinical_data = lung.get_clinical()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3e3bd2",
   "metadata": {},
   "source": [
    "**Exercise**. Examine these dataframes. In particular:\n",
    "\n",
    "- What do the values in the `rna_data` data frame represent?\n",
    "- Compare the dimensions of the `protein_data`, `rna_data`, and `clinical_data`. Are they equal?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a54f52a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample.IDs', 'Sample_Tumor_Normal', 'Smoking.Status', 'Stage',\n",
       "       'Region.of.Origin', 'Country.of.Origin', 'Age', 'Gender', 'Ethnicity',\n",
       "       'Height.cm', 'Weight.kg', 'BMI', 'Cigarettes.per.Day',\n",
       "       'Pack.Years.Smoked', 'Smoking.History', 'Secondhand.Smoke'],\n",
       "      dtype='object', name='Name')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore the data frames!\n",
    "\n",
    "rna_data.index # patient IDs\n",
    "rna_data.columns # RNA name\n",
    "\n",
    "rna_data.shape # 106 rows and 13482 columns\n",
    "clinical_data.shape # 210 rows and 17 columns\n",
    "clinical_data.columns\n",
    "# The rows and columns are not equal, some patients are included in some datasets and not in others"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cac2236",
   "metadata": {},
   "source": [
    "We can use the `intersect1d()` function from `numpy` to see which patient barcodes are shared between the data frames. **The reason we might be interested in these patients is that we can compare the transcriptomics to proteomics data.**\n",
    "\n",
    "1. Use `intersect1d()` to create a list that contains the patient IDs for each pair of data frames as mentioned below.\n",
    "2. Use a for loop to print the lengths of these three data frames, and compare them to the number of rows. The for loop skeleton is below. What do you notice about the three lengths? Do the numbers correspond to anything in particular?\n",
    "3. Finally, identify which intersection contains the patient IDs where there is data for all three levels of data (proteomics, transcriptomics, and clinical). How can we tell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "chemical-location",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of 0 is 211\n",
      "The length of 1 is 211\n",
      "The length of 2 is 211\n"
     ]
    }
   ],
   "source": [
    "# 1.\n",
    "name_intersects = [\n",
    "    np.intersect1d(protein_data.index , rna_data.index), # 0. fill in intersecting for protein/rna here\n",
    "    np.intersect1d(protein_data.index , clinical_data.index), # 1. fill in intersecting for protein/clinical here\n",
    "    np.intersect1d(rna_data.index , clinical_data.index), # 2. fill in the intersecting for rna/clinical here\n",
    "]\n",
    "#print(rna_data.index)\n",
    "\n",
    "#print(name_intersects)\n",
    "# 2. Print the lengths here\n",
    "# write a for loop here\n",
    "for i, intersects in enumerate(name_intersects):\n",
    "    print(f\"The length of {i} is {len(intersects)}\")  # fill in here\n",
    "# The length of the intersection between protein and clinical data match the number of rows in protein data.\n",
    "# The length of the intersection between rna and clinical data match the number of rows in the rna data. \n",
    "    \n",
    "# 3. Which comparison contains the patient names that have all three levels?\n",
    "# intersection of protein and rna data include all three levels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46124874",
   "metadata": {},
   "source": [
    "### Examining the clinical data\n",
    "Let's explore the clinical data in more detail.\n",
    "\n",
    "**Exercise**. First, let's use `head()` to glance as to what data is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5b38bbe6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Sample.IDs</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>Smoking.Status</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Region.of.Origin</th>\n",
       "      <th>Country.of.Origin</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Height.cm</th>\n",
       "      <th>Weight.kg</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Cigarettes.per.Day</th>\n",
       "      <th>Pack.Years.Smoked</th>\n",
       "      <th>Smoking.History</th>\n",
       "      <th>Secondhand.Smoke</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00001</th>\n",
       "      <td>C3L.00001</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>non-smoker</td>\n",
       "      <td>2A</td>\n",
       "      <td>western</td>\n",
       "      <td>usa</td>\n",
       "      <td>61.0</td>\n",
       "      <td>female</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>180.3</td>\n",
       "      <td>71.6</td>\n",
       "      <td>22.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lifelong non-smoker: Less than 100 cigarettes ...</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00009</th>\n",
       "      <td>C3L.00009</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>smoker</td>\n",
       "      <td>1A</td>\n",
       "      <td>western</td>\n",
       "      <td>usa</td>\n",
       "      <td>76.0</td>\n",
       "      <td>male</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>172.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>28.67</td>\n",
       "      <td>20.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00080</th>\n",
       "      <td>C3L.00080</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>smoker</td>\n",
       "      <td>1B</td>\n",
       "      <td>western</td>\n",
       "      <td>other</td>\n",
       "      <td>58.0</td>\n",
       "      <td>male</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>170.0</td>\n",
       "      <td>88.5</td>\n",
       "      <td>30.62</td>\n",
       "      <td>20.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>Current smoker: Includes daily and non-daily s...</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00083</th>\n",
       "      <td>C3L.00083</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1B</td>\n",
       "      <td>western</td>\n",
       "      <td>usa</td>\n",
       "      <td>71.0</td>\n",
       "      <td>male</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>177.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>29.37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Smoking history not available</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00093</th>\n",
       "      <td>C3L.00093</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>non-smoker</td>\n",
       "      <td>1B</td>\n",
       "      <td>western</td>\n",
       "      <td>usa</td>\n",
       "      <td>66.0</td>\n",
       "      <td>female</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>157.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>34.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lifelong non-smoker: Less than 100 cigarettes ...</td>\n",
       "      <td>Exposure to secondhand smoke history not avail...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Sample.IDs Sample_Tumor_Normal Smoking.Status Stage  \\\n",
       "Patient_ID                                                       \n",
       "C3L-00001   C3L.00001               Tumor     non-smoker    2A   \n",
       "C3L-00009   C3L.00009               Tumor         smoker    1A   \n",
       "C3L-00080   C3L.00080               Tumor         smoker    1B   \n",
       "C3L-00083   C3L.00083               Tumor            NaN    1B   \n",
       "C3L-00093   C3L.00093               Tumor     non-smoker    1B   \n",
       "\n",
       "Name       Region.of.Origin Country.of.Origin   Age  Gender  Ethnicity  \\\n",
       "Patient_ID                                                               \n",
       "C3L-00001           western               usa  61.0  female  caucasian   \n",
       "C3L-00009           western               usa  76.0    male  caucasian   \n",
       "C3L-00080           western             other  58.0    male  caucasian   \n",
       "C3L-00083           western               usa  71.0    male  caucasian   \n",
       "C3L-00093           western               usa  66.0  female  caucasian   \n",
       "\n",
       "Name        Height.cm  Weight.kg    BMI  Cigarettes.per.Day  \\\n",
       "Patient_ID                                                    \n",
       "C3L-00001       180.3       71.6  22.03                 NaN   \n",
       "C3L-00009       172.0       85.0  28.67                20.0   \n",
       "C3L-00080       170.0       88.5  30.62                20.0   \n",
       "C3L-00083       177.0       92.0  29.37                 NaN   \n",
       "C3L-00093       157.0       84.0  34.08                 NaN   \n",
       "\n",
       "Name        Pack.Years.Smoked  \\\n",
       "Patient_ID                      \n",
       "C3L-00001                 NaN   \n",
       "C3L-00009                62.0   \n",
       "C3L-00080                40.0   \n",
       "C3L-00083                 NaN   \n",
       "C3L-00093                 NaN   \n",
       "\n",
       "Name                                          Smoking.History  \\\n",
       "Patient_ID                                                      \n",
       "C3L-00001   Lifelong non-smoker: Less than 100 cigarettes ...   \n",
       "C3L-00009   Current smoker: Includes daily and non-daily s...   \n",
       "C3L-00080   Current smoker: Includes daily and non-daily s...   \n",
       "C3L-00083                       Smoking history not available   \n",
       "C3L-00093   Lifelong non-smoker: Less than 100 cigarettes ...   \n",
       "\n",
       "Name                                         Secondhand.Smoke  \n",
       "Patient_ID                                                     \n",
       "C3L-00001                                                 Yes  \n",
       "C3L-00009   Exposure to secondhand smoke history not avail...  \n",
       "C3L-00080   Exposure to secondhand smoke history not avail...  \n",
       "C3L-00083   Exposure to secondhand smoke history not avail...  \n",
       "C3L-00093   Exposure to secondhand smoke history not avail...  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# call head on the clinical data!\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e5486b7",
   "metadata": {},
   "source": [
    "**Exercise**. You might notice that the Age column values don't look like years -- they're in months. Divide the values by 12 in the `clinical_data` data frame to turn them into years. Then, use `head()` again to make sure the ages were converted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "61978ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>Age</th>\n",
       "      <th>CEA</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Lymphatic_Invasion</th>\n",
       "      <th>Mucinous</th>\n",
       "      <th>Perineural_Invasion</th>\n",
       "      <th>Polyps_History</th>\n",
       "      <th>Polyps_Present</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Subsite</th>\n",
       "      <th>Synchronous_Tumors</th>\n",
       "      <th>Tumor.Status</th>\n",
       "      <th>Vascular_Invasion</th>\n",
       "      <th>Vital.Status</th>\n",
       "      <th>pathalogy_N_stage</th>\n",
       "      <th>pathalogy_T_stage</th>\n",
       "      <th>age_category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>60.750000</td>\n",
       "      <td>4.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "      <td>Old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>69.833333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "      <td>Old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>75.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>Yes</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "      <td>Old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>54.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "      <td>Old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>57.916667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T2</td>\n",
       "      <td>Old</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Sample_Tumor_Normal        Age  CEA  Gender Lymphatic_Invasion  \\\n",
       "Patient_ID                                                                  \n",
       "01CO001                  Tumor  60.750000  4.9    Male                Yes   \n",
       "01CO005                  Tumor  69.833333  1.0  Female                 No   \n",
       "01CO006                  Tumor  75.333333  NaN  Female                Yes   \n",
       "01CO008                  Tumor  54.333333  NaN  Female                 No   \n",
       "01CO013                  Tumor  57.916667  NaN    Male                 No   \n",
       "\n",
       "Name            Mucinous Perineural_Invasion Polyps_History Polyps_Present  \\\n",
       "Patient_ID                                                                   \n",
       "01CO001         Mucinous                  No             No             No   \n",
       "01CO005     Not Mucinous                  No            Yes            Yes   \n",
       "01CO006         Mucinous                 Yes             No             No   \n",
       "01CO008         Mucinous                  No             No            Yes   \n",
       "01CO013     Not Mucinous                  No            Yes             No   \n",
       "\n",
       "Name            Stage           Subsite Synchronous_Tumors Tumor.Status  \\\n",
       "Patient_ID                                                                \n",
       "01CO001     Stage III     Sigmoid Colon                 No   Tumor free   \n",
       "01CO005      Stage II     Sigmoid Colon                 No   Tumor free   \n",
       "01CO006     Stage III   Ascending Colon                Yes   With tumor   \n",
       "01CO008      Stage II  Descending Colon                 No   With tumor   \n",
       "01CO013       Stage I     Sigmoid Colon                 No   Tumor free   \n",
       "\n",
       "Name       Vascular_Invasion Vital.Status pathalogy_N_stage pathalogy_T_stage  \\\n",
       "Patient_ID                                                                      \n",
       "01CO001                  Yes       Living               N2b               T4a   \n",
       "01CO005                   No     Deceased                N0                T3   \n",
       "01CO006                  Yes       Living               N2b               T4a   \n",
       "01CO008                   No       Living                N0                T3   \n",
       "01CO013                   No       Living                N0                T2   \n",
       "\n",
       "Name       age_category  \n",
       "Patient_ID               \n",
       "01CO001             Old  \n",
       "01CO005             Old  \n",
       "01CO006             Old  \n",
       "01CO008             Old  \n",
       "01CO013             Old  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert ages here!\n",
    "#clinical_data[\"Age\"] /= 12\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d704ebf5",
   "metadata": {},
   "source": [
    "There are actually control (non-cancer) tissue samples in the data! Let's show this using the `unique()` function from numpy.\n",
    "\n",
    "**Exercise.** Call the `unique()` function on the `Sample_Tumor_Normal` column. The function will return the unique values in the column in a sorted order -- this is super useful for examining categorical variables, like tumor stage and tumor status, for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "416c93c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Normal', 'Tumor'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the levels of the Sample_Tumor_Normal column with unique()\n",
    "np.unique(clinical_data[\"Sample_Tumor_Normal\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5ff472",
   "metadata": {},
   "source": [
    "**Exercise.** Let's examine how tumor stage varies as a function of age:\n",
    "\n",
    "1. Import the plotting libraries `matplotlib.pyplot` and `seaborn` (use the standard abbreviations presented last time).\n",
    "2. Create an `age_category` column in `clinical_data`. As before, we'll define \"Young\" as under 50, and \"Old\" as 50 and older. (Hint: use the `where()` function from numpy, the equivalent to `ifelse()`).\n",
    "3.  The \"Stage\" information for these patients is NaN (the pandas version of NA). Use the `isna()` function from pandas and boolean indexing (i.e. `where()` from numpy again) to replace these NaN values in the `Stage` column with \"No cancer\".\n",
    "3. Using nested for loop and boolean indexing, count the number of old and young patients that have No cancer, Stage I, Stage II, Stage III, and Stage IV cancer. \n",
    "4. Compare your results to the barplot. Do your numbers make sense?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6e84c387",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient_ID\n",
      "C3L-00001       2A\n",
      "C3L-00009       1A\n",
      "C3L-00080       1B\n",
      "C3L-00083       1B\n",
      "C3L-00093       1B\n",
      "              ... \n",
      "C3N-02582.N    NaN\n",
      "C3N-02586.N    NaN\n",
      "C3N-02587.N    NaN\n",
      "C3N-02588.N    NaN\n",
      "C3N-02729.N    NaN\n",
      "Name: Stage, Length: 211, dtype: object\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/g9/8xmnj2215vx2gnwxq0l4cn600000gn/T/ipykernel_44340/2331335177.py:24: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  assert(np.all(stage_categories == ['No cancer', 'Stage I', 'Stage II', 'Stage III', 'Stage IV']))\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/g9/8xmnj2215vx2gnwxq0l4cn600000gn/T/ipykernel_44340/2331335177.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;31m## this will programatically get all the stages in order\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0mstage_categories\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstage_categories\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m \u001b[0;32massert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstage_categories\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'No cancer'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Stage I'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Stage II'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Stage III'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Stage IV'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;31m# 4. Loop through all the stage categories and count\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# 1. Import libraries here\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# 2. Create the age_category column in clinical_data\n",
    "clinical_data[\"age_category\"] = np.where(\n",
    "    clinical_data[\"Age\"] < 50,\n",
    "    \"Young\",\n",
    "    \"Old\"\n",
    ")\n",
    "\n",
    "print(clinical_data[\"Stage\"])\n",
    "\n",
    "# 3. Replace NaN with No cancer\n",
    "stage_categories = np.where(\n",
    "    clinical_data[\"Stage\"].isna(),\n",
    "    \"No cancer\",\n",
    "    clinical_data[\"Stage\"]\n",
    ")\n",
    "\n",
    "\n",
    "## this will programatically get all the stages in order\n",
    "stage_categories = np.unique(stage_categories)\n",
    "assert(np.all(stage_categories == ['No cancer', 'Stage I', 'Stage II', 'Stage III', 'Stage IV']))\n",
    "\n",
    "# 4. Loop through all the stage categories and count\n",
    "\n",
    "for age_cat in [\"Old\", \"Young\"]: # first get the old data, then the young data\n",
    "    print(age_cat)\n",
    "    # fill this in with the cancer category:\n",
    "    for stage_cat in stage_categories:\n",
    "    #    # 1. create a subset of the data frame with the old/young patients\n",
    "        # 2. count the number of patients with the stage of cancer and print\n",
    "        subset = clinical_data[clinical_data[\"age_category\"] == age_cat]\n",
    "        count = subset[subset[\"Stage\"] == stage_cat]\n",
    "        print(stage_cat, len(count))\n",
    "\n",
    "\n",
    "        \n",
    "# 5. A sample barplot to compare your results\n",
    "clinical_data_copy = clinical_data\n",
    "clinical_data_copy[\"Stage\"] = np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "sns.countplot(\n",
    "    x = \"age_category\",\n",
    "    hue = \"Stage\",\n",
    "    hue_order = [\"No cancer\", \"Stage I\", \"Stage II\", \"Stage III\", \"Stage IV\"],\n",
    "    data = clinical_data_copy\n",
    ")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# yes, my results from part 4 match the bar lengths in part 5."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4747fb4c",
   "metadata": {},
   "source": [
    "## Comparing Transcriptomic and Proteomic Data\n",
    "A natural question to ask is if the levels of RNA expression correlate with protein expression. We can explore this question with the help of the `scipy` library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c5e207ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats # we are using the stats package in particular"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1395e3d7",
   "metadata": {},
   "source": [
    "The first thing we need to do is to identify which patients and genes are shared between the transcriptomic and proteomic data sets. We need to identify which genes are shared, but luckily, we already got that data before for the patients! We will use this data to create two new data frames that contain only the rows and column names that are shared between these two data sets. \n",
    "\n",
    "**Exercise**. To make sure the data frames match:\n",
    "1. Identify the names of the genes that are shared between the two datasets (refer to the exercise when we created the dictionary). Is this data the row or column names?\n",
    "2. Create the `rna_shared` and `prot_shared` data frames as described above (Hint: how can we access rows/columns by name?).\n",
    "3. Finally, we need to make sure all the rows and columns are in the correct order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e2589e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Identify the genes (RNA, protein) shared between the two data sets \n",
    "shared_rna_prot = np.intersect1d(rna_data.columns, new_protein_columns)\n",
    "\n",
    "# 2. Create the two data frames\n",
    "rna_shared = rna_data.loc[name_intersects[0], shared_rna_prot]\n",
    "prot_shared = protein_data.loc[name_intersects[0], shared_rna_prot]\n",
    "\n",
    "# 3. make sure that everything is in the correct order\n",
    "#assert(np.all(rna_shared.index == prot_shared.index))\n",
    "#assert(np.all(rna_shared.columns == prot_shared.columns))\n",
    "\n",
    "# you can, in fact, chain comparisons in python\n",
    "# DO NOT DO THIS IN ANY OTHER LANGUAGE\n",
    "#assert(rna_shared.shape == prot_shared.shape == (96, 7572))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "769b1472",
   "metadata": {},
   "source": [
    "Now, we can see how correlated the RNA and protein levels are. We'll use Spearman correlation from the `stats` library, which is `spearmanr()`. For example, to get the correlation in the KRAS gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0d7aff46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The correlation is 0.078 (p = 0.461116).\n"
     ]
    }
   ],
   "source": [
    "# we need the nan_policy=\"omit\" to throw out NaN values\n",
    "corr, pval = stats.spearmanr(\n",
    "    rna_shared[\"CYP\"], \n",
    "    prot_shared[\"KRAS\"], \n",
    "    nan_policy=\"omit\")\n",
    "\n",
    "# round() to truncate the numbers\n",
    "print(f\"The correlation is {round(corr, 3)} (p = {round(pval, 6)}).\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3174183c",
   "metadata": {},
   "source": [
    "**Exercise**. Investigate the correlation between the RNA and the protein data for the CD8A gene, printing the values. Then, using seaborn, create a scatterplot, with RNA on the x-axis and protein on the y-axis (be sure to create your fig/ax!). Does the correlation look like it makes sense? As usual, be sure to label the axes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a1002dc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The correlation is 0.863 (p = 0.0).\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Data must be 1-dimensional",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/g9/8xmnj2215vx2gnwxq0l4cn600000gn/T/ipykernel_6303/198778341.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# create scatter plot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m sns.scatterplot(\n\u001b[0m\u001b[1;32m     17\u001b[0m     \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrna_shared\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprot_shared\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/_decorators.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m             )\n\u001b[1;32m     45\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/relational.py\u001b[0m in \u001b[0;36mscatterplot\u001b[0;34m(x, y, hue, style, size, data, palette, hue_order, hue_norm, sizes, size_order, size_norm, markers, style_order, x_bins, y_bins, units, estimator, ci, n_boot, alpha, x_jitter, y_jitter, legend, ax, **kwargs)\u001b[0m\n\u001b[1;32m    806\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    807\u001b[0m     \u001b[0mvariables\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_ScatterPlotter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_semantics\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlocals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 808\u001b[0;31m     p = _ScatterPlotter(\n\u001b[0m\u001b[1;32m    809\u001b[0m         \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvariables\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    810\u001b[0m         \u001b[0mx_bins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mx_bins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_bins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0my_bins\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/relational.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, variables, x_bins, y_bins, estimator, ci, n_boot, alpha, x_jitter, y_jitter, legend)\u001b[0m\n\u001b[1;32m    585\u001b[0m         )\n\u001b[1;32m    586\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 587\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    588\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    589\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malpha\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malpha\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/_core.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, variables)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0massign_variables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mvar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcls\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_semantic_mappings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/_core.py\u001b[0m in \u001b[0;36massign_variables\u001b[0;34m(self, data, variables)\u001b[0m\n\u001b[1;32m    666\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    667\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minput_format\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"long\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 668\u001b[0;31m             plot_data, variables = self._assign_variables_longform(\n\u001b[0m\u001b[1;32m    669\u001b[0m                 \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mvariables\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    670\u001b[0m             )\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/seaborn/_core.py\u001b[0m in \u001b[0;36m_assign_variables_longform\u001b[0;34m(self, data, **kwargs)\u001b[0m\n\u001b[1;32m    925\u001b[0m         \u001b[0;31m# Construct a tidy plot DataFrame. This will convert a number of\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    926\u001b[0m         \u001b[0;31m# types automatically, aligning on index in case of pandas objects\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 927\u001b[0;31m         \u001b[0mplot_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mplot_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    928\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    929\u001b[0m         \u001b[0;31m# Reduce the variables dictionary to fields with valid data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    634\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    635\u001b[0m             \u001b[0;31m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 636\u001b[0;31m             \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict_to_mgr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtyp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmanager\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    637\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMaskedArray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    638\u001b[0m             \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmrecords\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mmrecords\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36mdict_to_mgr\u001b[0;34m(data, index, columns, dtype, typ, copy)\u001b[0m\n\u001b[1;32m    500\u001b[0m         \u001b[0;31m# TODO: can we get rid of the dt64tz special case above?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    501\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 502\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0marrays_to_mgr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtyp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtyp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsolidate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    503\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36marrays_to_mgr\u001b[0;34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[0m\n\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m         \u001b[0;31m# don't force copy because getting jammed in an ndarray anyway\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 125\u001b[0;31m         \u001b[0marrays\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_homogenize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    126\u001b[0m         \u001b[0;31m# _homogenize ensures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m         \u001b[0;31m#  - all(len(x) == len(index) for x in arrays)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36m_homogenize\u001b[0;34m(data, index, dtype)\u001b[0m\n\u001b[1;32m    623\u001b[0m                 \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfast_multiget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdefault\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnan\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 625\u001b[0;31m             val = sanitize_array(\n\u001b[0m\u001b[1;32m    626\u001b[0m                 \u001b[0mval\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_cast_failure\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m             )\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/construction.py\u001b[0m in \u001b[0;36msanitize_array\u001b[0;34m(data, index, dtype, copy, raise_cast_failure, allow_2d)\u001b[0m\n\u001b[1;32m    596\u001b[0m                 \u001b[0msubarr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmaybe_infer_to_datetimelike\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    597\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 598\u001b[0;31m     \u001b[0msubarr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_sanitize_ndim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_2d\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_2d\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    599\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    600\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/construction.py\u001b[0m in \u001b[0;36m_sanitize_ndim\u001b[0;34m(result, data, dtype, index, allow_2d)\u001b[0m\n\u001b[1;32m    647\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mallow_2d\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    648\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 649\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Data must be 1-dimensional\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    650\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mExtensionDtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    651\u001b[0m             \u001b[0;31m# i.e. PandasDtype(\"O\")\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Data must be 1-dimensional"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD8CAYAAAB0IB+mAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAANQklEQVR4nO3cX4il9X3H8fenuxEak0aJk5DurmRb1pi90KITI6VpTUObXXuxBLxQQ6QSWKQx5FIpNLnwprkohKBmWWSR3GQvGkk2ZRMplMSCNd1Z8N8qynSlOl3BNYYUDFRWv704p51hnHWenXNmZp3v+wUD85znNzPf+TH73mfPznlSVUiStr7f2ewBJEkbw+BLUhMGX5KaMPiS1ITBl6QmDL4kNbFq8JMcSfJakmfPcz5JvptkPsnTSa6b/piSpEkNucJ/GNj3Huf3A3vGbweB700+liRp2lYNflU9BrzxHksOAN+vkSeAy5J8YloDSpKmY/sUPscO4JUlxwvjx15dvjDJQUb/CuDSSy+9/uqrr57Cl5ekPk6ePPl6Vc2s5WOnEfys8NiK92uoqsPAYYDZ2dmam5ubwpeXpD6S/OdaP3Yav6WzAOxacrwTODOFzytJmqJpBP8YcMf4t3VuBH5TVe96OkeStLlWfUonyQ+Am4ArkiwA3wI+AFBVh4DjwM3APPBb4M71GlaStHarBr+qblvlfAFfm9pEkqR14SttJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJamJQ8JPsS/JCkvkk965w/iNJfpLkqSSnktw5/VElSZNYNfhJtgEPAPuBvcBtSfYuW/Y14Lmquha4CfiHJJdMeVZJ0gSGXOHfAMxX1emqegs4ChxYtqaADycJ8CHgDeDcVCeVJE1kSPB3AK8sOV4YP7bU/cCngTPAM8A3quqd5Z8oycEkc0nmzp49u8aRJUlrMST4WeGxWnb8ReBJ4PeBPwLuT/J77/qgqsNVNVtVszMzMxc4qiRpEkOCvwDsWnK8k9GV/FJ3Ao/UyDzwEnD1dEaUJE3DkOCfAPYk2T3+j9hbgWPL1rwMfAEgyceBTwGnpzmoJGky21dbUFXnktwNPApsA45U1akkd43PHwLuAx5O8gyjp4DuqarX13FuSdIFWjX4AFV1HDi+7LFDS94/A/zldEeTJE2Tr7SVpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDUxKPhJ9iV5Icl8knvPs+amJE8mOZXkF9MdU5I0qe2rLUiyDXgA+AtgATiR5FhVPbdkzWXAg8C+qno5ycfWaV5J0hoNucK/AZivqtNV9RZwFDiwbM3twCNV9TJAVb023TElSZMaEvwdwCtLjhfGjy11FXB5kp8nOZnkjpU+UZKDSeaSzJ09e3ZtE0uS1mRI8LPCY7XseDtwPfBXwBeBv0ty1bs+qOpwVc1W1ezMzMwFDytJWrtVn8NndEW/a8nxTuDMCmter6o3gTeTPAZcC7w4lSklSRMbcoV/AtiTZHeSS4BbgWPL1vwY+FyS7Uk+CHwWeH66o0qSJrHqFX5VnUtyN/AosA04UlWnktw1Pn+oqp5P8jPgaeAd4KGqenY9B5ckXZhULX86fmPMzs7W3NzcpnxtSXq/SnKyqmbX8rG+0laSmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmBgU/yb4kLySZT3Lve6z7TJK3k9wyvRElSdOwavCTbAMeAPYDe4Hbkuw9z7pvA49Oe0hJ0uSGXOHfAMxX1emqegs4ChxYYd3XgR8Cr01xPknSlAwJ/g7glSXHC+PH/l+SHcCXgEPv9YmSHEwyl2Tu7NmzFzqrJGkCQ4KfFR6rZcffAe6pqrff6xNV1eGqmq2q2ZmZmYEjSpKmYfuANQvAriXHO4Ezy9bMAkeTAFwB3JzkXFX9aBpDSpImNyT4J4A9SXYD/wXcCty+dEFV7f6/95M8DPyTsZeki8uqwa+qc0nuZvTbN9uAI1V1Ksld4/Pv+by9JOniMOQKn6o6Dhxf9tiKoa+qv558LEnStPlKW0lqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSE4OCn2RfkheSzCe5d4XzX07y9Pjt8STXTn9USdIkVg1+km3AA8B+YC9wW5K9y5a9BPxZVV0D3AccnvagkqTJDLnCvwGYr6rTVfUWcBQ4sHRBVT1eVb8eHz4B7JzumJKkSQ0J/g7glSXHC+PHzuerwE9XOpHkYJK5JHNnz54dPqUkaWJDgp8VHqsVFyafZxT8e1Y6X1WHq2q2qmZnZmaGTylJmtj2AWsWgF1LjncCZ5YvSnIN8BCwv6p+NZ3xJEnTMuQK/wSwJ8nuJJcAtwLHli5IciXwCPCVqnpx+mNKkia16hV+VZ1LcjfwKLANOFJVp5LcNT5/CPgm8FHgwSQA56pqdv3GliRdqFSt+HT8upudna25ublN+dqS9H6V5ORaL6h9pa0kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNDAp+kn1JXkgyn+TeFc4nyXfH559Oct30R5UkTWLV4CfZBjwA7Af2Arcl2bts2X5gz/jtIPC9Kc8pSZrQkCv8G4D5qjpdVW8BR4EDy9YcAL5fI08AlyX5xJRnlSRNYPuANTuAV5YcLwCfHbBmB/Dq0kVJDjL6FwDA/yR59oKm3bquAF7f7CEuEu7FIvdikXux6FNr/cAhwc8Kj9Ua1lBVh4HDAEnmqmp2wNff8tyLRe7FIvdikXuxKMncWj92yFM6C8CuJcc7gTNrWCNJ2kRDgn8C2JNkd5JLgFuBY8vWHAPuGP+2zo3Ab6rq1eWfSJK0eVZ9SqeqziW5G3gU2AYcqapTSe4anz8EHAduBuaB3wJ3Dvjah9c89dbjXixyLxa5F4vci0Vr3otUveupdknSFuQrbSWpCYMvSU2se/C9LcOiAXvx5fEePJ3k8STXbsacG2G1vViy7jNJ3k5yy0bOt5GG7EWSm5I8meRUkl9s9IwbZcCfkY8k+UmSp8Z7MeT/C993khxJ8tr5Xqu05m5W1bq9MfpP3v8A/gC4BHgK2Ltszc3ATxn9Lv+NwC/Xc6bNehu4F38MXD5+f3/nvViy7l8Y/VLALZs99yb+XFwGPAdcOT7+2GbPvYl78bfAt8fvzwBvAJds9uzrsBd/ClwHPHue82vq5npf4XtbhkWr7kVVPV5Vvx4fPsHo9Qxb0ZCfC4CvAz8EXtvI4TbYkL24HXikql4GqKqtuh9D9qKADycJ8CFGwT+3sWOuv6p6jNH3dj5r6uZ6B/98t1y40DVbwYV+n19l9Df4VrTqXiTZAXwJOLSBc22GIT8XVwGXJ/l5kpNJ7tiw6TbWkL24H/g0oxd2PgN8o6re2ZjxLipr6uaQWytMYmq3ZdgCBn+fST7PKPh/sq4TbZ4he/Ed4J6qent0MbdlDdmL7cD1wBeA3wX+LckTVfXieg+3wYbsxReBJ4E/B/4Q+Ock/1pV/73Os11s1tTN9Q6+t2VYNOj7THIN8BCwv6p+tUGzbbQhezELHB3H/grg5iTnqupHGzLhxhn6Z+T1qnoTeDPJY8C1wFYL/pC9uBP4+xo9kT2f5CXgauDfN2bEi8aaurneT+l4W4ZFq+5FkiuBR4CvbMGrt6VW3Yuq2l1Vn6yqTwL/CPzNFow9DPsz8mPgc0m2J/kgo7vVPr/Bc26EIXvxMqN/6ZDk44zuHHl6Q6e8OKypm+t6hV/rd1uG952Be/FN4KPAg+Mr23O1Be8QOHAvWhiyF1X1fJKfAU8D7wAPVdWWu7X4wJ+L+4CHkzzD6GmNe6pqy902OckPgJuAK5IsAN8CPgCTddNbK0hSE77SVpKaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrifwHXe3WluIZOawAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "gene = \"GSTM1\" # this will save you some typing and make the code more flexible for exploring other genes\n",
    "\n",
    "# get correlation and print\n",
    "corr, pval = stats.spearmanr(\n",
    "    rna_shared[gene], \n",
    "    prot_shared[gene], \n",
    "    nan_policy=\"omit\")\n",
    "\n",
    "print(f\"The correlation is {round(corr, 3)} (p = {round(pval, 6)}).\")\n",
    "\n",
    "# create scatter plot\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sns.scatterplot(\n",
    "    x = rna_shared[gene], \n",
    "    y = prot_shared[gene], \n",
    "    ax = ax\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "556dffec",
   "metadata": {},
   "source": [
    "## Heatmaps\n",
    "A problem arises if we want to compare many correlations. Heatmaps are useful for visualizing a large number of comparisons. To make a heatmap, we'll use the `heatmap()` function from seaborn (the `cluster()` function is closely related; the difference is that `clustermap()` will reorganize the rows/columns so that similar rows/columns are grouped near each other). Let's do the following: \n",
    "\n",
    "1. Set up our data frame to hold all comparisons. All you need to do is to access the first 20 gene names.\n",
    "2. Calculate the correlations for the first 20 genes (just to save time). You'll need to use two for loops.\n",
    "3. Call `clustermap()` -- this is filled in for you!\n",
    "4. Interpret the data. There is a \"light\" diagonal along the heatmap from the top-left to bottom-right. Is this expected? Why or why not?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a95838ba",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rna_shared' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/g9/8xmnj2215vx2gnwxq0l4cn600000gn/T/ipykernel_68759/2686268085.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mncomparisons\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m20\u001b[0m \u001b[0;31m# define this variable in case we want to change the number of correlations to test\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m                   \u001b[0;31m# this makes it less likely you'll forget to change a number, e.g. in the data frame shape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mgene_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrna_shared\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mncomparisons\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# get the first ncomparisons gene names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rna_shared' is not defined"
     ]
    }
   ],
   "source": [
    "ncomparisons = 20 # define this variable in case we want to change the number of correlations to test\n",
    "                  # this makes it less likely you'll forget to change a number, e.g. in the data frame shape\n",
    "gene_names = rna_shared.columns[:ncomparisons] # get the first ncomparisons gene names\n",
    "print(gene_names)\n",
    "\n",
    "\n",
    "# Don't worry about this code\n",
    "# It's good practice to declare your data frame beforehand (it's much faster than appending to a list)\n",
    "# We fill everything in with 0 just as a placeholder\n",
    "corr_df = pd.DataFrame(np.ndarray(shape=(ncomparisons, ncomparisons), dtype=np.float16),\n",
    "                      index = gene_names,\n",
    "                      columns = gene_names)\n",
    "\n",
    "# 2. fill in the data frame!\n",
    "for g1 in gene_names:\n",
    "    for g2 in gene_names:\n",
    "        # calculate the correlations between protein and RNA\n",
    "        # then, use .loc[] to store the correlation in corr_df\n",
    "        corr, pval = stats.spearmanr(\n",
    "        rna_shared[g1], \n",
    "        prot_shared[g2], \n",
    "        nan_policy=\"omit\")\n",
    "        \n",
    "        corr_df.loc[g1,g2] = corr\n",
    "\n",
    "# 3. create the heat map\n",
    "sns.heatmap(\n",
    "    corr_df,\n",
    "    cmap='mako'\n",
    ")\n",
    "\n",
    "plt.show()\n",
    "plt.savefig('heatmap.png', bbox_inches='tight')\n",
    "# 4. interpret!\n",
    "# The light diagonal line shows the correlation between the same gene's rna and protein data. Light means that it is \n",
    "# strongly correlated and positive. This makes sense because the same gene's rna and protein data should have a direct\n",
    "# correlation. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "attached-charity",
   "metadata": {},
   "source": [
    "### Saving pandas dataframes\n",
    "While you can certainly reload in the data every time you wish to use it, we also want to make sure you know how to save and read in a pandas dataframes. We will utilize the `to_csv()` function and `read_csv()` function. Notice in the `read_csv()` function we utilize the `index_col` argument. To observe what this argument does, try removing it and looking at the dataframe. Notice what the index values are. \n",
    "\n",
    "**Exercise**. To save our data:\n",
    "1. Make a folder in your `analysis_data` folder called `cptac`.\n",
    "2. Edit the path to save into `analysis_data`\n",
    "3. Also save the RNA and clinical data as well to the same directory (with the same name convention)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "resistant-approach",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADACP1</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAED1</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>10.1984</td>\n",
       "      <td>13.8117</td>\n",
       "      <td>7.0224</td>\n",
       "      <td>9.4115</td>\n",
       "      <td>9.7432</td>\n",
       "      <td>4.5850</td>\n",
       "      <td>8.4635</td>\n",
       "      <td>7.6221</td>\n",
       "      <td>11.2336</td>\n",
       "      <td>8.3219</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1046</td>\n",
       "      <td>10.5765</td>\n",
       "      <td>10.5294</td>\n",
       "      <td>7.0553</td>\n",
       "      <td>8.4263</td>\n",
       "      <td>8.6865</td>\n",
       "      <td>10.0307</td>\n",
       "      <td>9.1699</td>\n",
       "      <td>10.2784</td>\n",
       "      <td>10.4263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>10.1586</td>\n",
       "      <td>12.6434</td>\n",
       "      <td>6.7279</td>\n",
       "      <td>8.9972</td>\n",
       "      <td>9.9084</td>\n",
       "      <td>4.7004</td>\n",
       "      <td>8.2046</td>\n",
       "      <td>7.0768</td>\n",
       "      <td>10.5362</td>\n",
       "      <td>9.6402</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6883</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>9.2808</td>\n",
       "      <td>7.1799</td>\n",
       "      <td>9.2946</td>\n",
       "      <td>9.1799</td>\n",
       "      <td>9.8408</td>\n",
       "      <td>10.5622</td>\n",
       "      <td>10.4512</td>\n",
       "      <td>10.4305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>9.2119</td>\n",
       "      <td>15.0040</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>9.7056</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>8.4676</td>\n",
       "      <td>7.1293</td>\n",
       "      <td>10.9337</td>\n",
       "      <td>9.5661</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8424</td>\n",
       "      <td>9.8611</td>\n",
       "      <td>10.2107</td>\n",
       "      <td>7.6582</td>\n",
       "      <td>8.7415</td>\n",
       "      <td>9.0741</td>\n",
       "      <td>10.3309</td>\n",
       "      <td>10.0512</td>\n",
       "      <td>10.2703</td>\n",
       "      <td>10.5430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>10.7322</td>\n",
       "      <td>13.9284</td>\n",
       "      <td>4.9542</td>\n",
       "      <td>9.1497</td>\n",
       "      <td>8.9425</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.4179</td>\n",
       "      <td>7.7074</td>\n",
       "      <td>10.7194</td>\n",
       "      <td>9.3376</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6812</td>\n",
       "      <td>9.0795</td>\n",
       "      <td>9.5118</td>\n",
       "      <td>8.9159</td>\n",
       "      <td>9.9542</td>\n",
       "      <td>10.0168</td>\n",
       "      <td>10.3365</td>\n",
       "      <td>9.9307</td>\n",
       "      <td>10.2538</td>\n",
       "      <td>10.8925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>9.7160</td>\n",
       "      <td>12.9963</td>\n",
       "      <td>6.3576</td>\n",
       "      <td>9.6202</td>\n",
       "      <td>9.2738</td>\n",
       "      <td>4.1699</td>\n",
       "      <td>7.0980</td>\n",
       "      <td>5.9069</td>\n",
       "      <td>10.2252</td>\n",
       "      <td>9.7944</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1344</td>\n",
       "      <td>9.3174</td>\n",
       "      <td>10.0238</td>\n",
       "      <td>7.4429</td>\n",
       "      <td>8.7381</td>\n",
       "      <td>9.6165</td>\n",
       "      <td>10.3848</td>\n",
       "      <td>11.3066</td>\n",
       "      <td>11.2174</td>\n",
       "      <td>10.7616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21CO007</th>\n",
       "      <td>10.0768</td>\n",
       "      <td>13.4762</td>\n",
       "      <td>6.4594</td>\n",
       "      <td>9.6018</td>\n",
       "      <td>9.1573</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>6.2095</td>\n",
       "      <td>10.9159</td>\n",
       "      <td>10.9315</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1293</td>\n",
       "      <td>9.5622</td>\n",
       "      <td>10.2958</td>\n",
       "      <td>7.5314</td>\n",
       "      <td>8.3707</td>\n",
       "      <td>8.8202</td>\n",
       "      <td>9.8234</td>\n",
       "      <td>9.4136</td>\n",
       "      <td>9.7245</td>\n",
       "      <td>10.6777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO004</th>\n",
       "      <td>10.9293</td>\n",
       "      <td>12.6131</td>\n",
       "      <td>5.5236</td>\n",
       "      <td>9.3332</td>\n",
       "      <td>9.2877</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.3663</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>10.9031</td>\n",
       "      <td>8.8361</td>\n",
       "      <td>...</td>\n",
       "      <td>9.9687</td>\n",
       "      <td>9.5736</td>\n",
       "      <td>10.5108</td>\n",
       "      <td>8.4346</td>\n",
       "      <td>9.0768</td>\n",
       "      <td>8.8857</td>\n",
       "      <td>10.2715</td>\n",
       "      <td>9.4737</td>\n",
       "      <td>9.6511</td>\n",
       "      <td>10.9114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO006</th>\n",
       "      <td>9.2432</td>\n",
       "      <td>14.2620</td>\n",
       "      <td>7.5546</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>8.7039</td>\n",
       "      <td>5.5546</td>\n",
       "      <td>7.3129</td>\n",
       "      <td>8.0389</td>\n",
       "      <td>11.1592</td>\n",
       "      <td>9.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8903</td>\n",
       "      <td>9.9972</td>\n",
       "      <td>10.1923</td>\n",
       "      <td>7.4998</td>\n",
       "      <td>8.1447</td>\n",
       "      <td>8.7582</td>\n",
       "      <td>10.1497</td>\n",
       "      <td>10.4062</td>\n",
       "      <td>10.2491</td>\n",
       "      <td>10.3902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24CO005</th>\n",
       "      <td>8.6546</td>\n",
       "      <td>13.0250</td>\n",
       "      <td>8.0498</td>\n",
       "      <td>10.2228</td>\n",
       "      <td>9.3729</td>\n",
       "      <td>3.1699</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>8.3399</td>\n",
       "      <td>11.0566</td>\n",
       "      <td>8.0980</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8933</td>\n",
       "      <td>10.2621</td>\n",
       "      <td>10.8727</td>\n",
       "      <td>6.2479</td>\n",
       "      <td>7.8392</td>\n",
       "      <td>8.1849</td>\n",
       "      <td>9.6917</td>\n",
       "      <td>8.3264</td>\n",
       "      <td>7.8704</td>\n",
       "      <td>10.3740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27CO004</th>\n",
       "      <td>9.8025</td>\n",
       "      <td>13.1139</td>\n",
       "      <td>7.9069</td>\n",
       "      <td>9.7582</td>\n",
       "      <td>9.2479</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.7682</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>11.0049</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6439</td>\n",
       "      <td>9.8857</td>\n",
       "      <td>10.3049</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>8.6110</td>\n",
       "      <td>9.1624</td>\n",
       "      <td>9.3106</td>\n",
       "      <td>11.0532</td>\n",
       "      <td>9.6777</td>\n",
       "      <td>9.6900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows Ã— 13482 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A1CF      A2M  A4GALT     AAAS    AACS  AADACP1   AADAT  \\\n",
       "Patient_ID                                                               \n",
       "01CO001     10.1984  13.8117  7.0224   9.4115  9.7432   4.5850  8.4635   \n",
       "01CO005     10.1586  12.6434  6.7279   8.9972  9.9084   4.7004  8.2046   \n",
       "01CO006      9.2119  15.0040  6.8580   9.7056  9.1997   5.0000  8.4676   \n",
       "01CO008     10.7322  13.9284  4.9542   9.1497  8.9425   6.0224  7.4179   \n",
       "01CO013      9.7160  12.9963  6.3576   9.6202  9.2738   4.1699  7.0980   \n",
       "...             ...      ...     ...      ...     ...      ...     ...   \n",
       "21CO007     10.0768  13.4762  6.4594   9.6018  9.1573   5.0000  7.6865   \n",
       "22CO004     10.9293  12.6131  5.5236   9.3332  9.2877   5.0000  7.3663   \n",
       "22CO006      9.2432  14.2620  7.5546   9.3264  8.7039   5.5546  7.3129   \n",
       "24CO005      8.6546  13.0250  8.0498  10.2228  9.3729   3.1699  7.0444   \n",
       "27CO004      9.8025  13.1139  7.9069   9.7582  9.2479   6.0224  7.7682   \n",
       "\n",
       "             AAED1    AAGAB     AAK1  ...     ZW10   ZWILCH    ZWINT    ZXDA  \\\n",
       "Patient_ID                            ...                                      \n",
       "01CO001     7.6221  11.2336   8.3219  ...  10.1046  10.5765  10.5294  7.0553   \n",
       "01CO005     7.0768  10.5362   9.6402  ...   9.6883   9.3264   9.2808  7.1799   \n",
       "01CO006     7.1293  10.9337   9.5661  ...   9.8424   9.8611  10.2107  7.6582   \n",
       "01CO008     7.7074  10.7194   9.3376  ...   9.6812   9.0795   9.5118  8.9159   \n",
       "01CO013     5.9069  10.2252   9.7944  ...  10.1344   9.3174  10.0238  7.4429   \n",
       "...            ...      ...      ...  ...      ...      ...      ...     ...   \n",
       "21CO007     6.2095  10.9159  10.9315  ...  10.1293   9.5622  10.2958  7.5314   \n",
       "22CO004     6.8580  10.9031   8.8361  ...   9.9687   9.5736  10.5108  8.4346   \n",
       "22CO006     8.0389  11.1592   9.2597  ...   9.8903   9.9972  10.1923  7.4998   \n",
       "24CO005     8.3399  11.0566   8.0980  ...   9.8933  10.2621  10.8727  6.2479   \n",
       "27CO004     7.0444  11.0049   9.1997  ...   9.6439   9.8857  10.3049  7.6865   \n",
       "\n",
       "              ZXDB     ZXDC   ZYG11B      ZYX    ZZEF1     ZZZ3  \n",
       "Patient_ID                                                       \n",
       "01CO001     8.4263   8.6865  10.0307   9.1699  10.2784  10.4263  \n",
       "01CO005     9.2946   9.1799   9.8408  10.5622  10.4512  10.4305  \n",
       "01CO006     8.7415   9.0741  10.3309  10.0512  10.2703  10.5430  \n",
       "01CO008     9.9542  10.0168  10.3365   9.9307  10.2538  10.8925  \n",
       "01CO013     8.7381   9.6165  10.3848  11.3066  11.2174  10.7616  \n",
       "...            ...      ...      ...      ...      ...      ...  \n",
       "21CO007     8.3707   8.8202   9.8234   9.4136   9.7245  10.6777  \n",
       "22CO004     9.0768   8.8857  10.2715   9.4737   9.6511  10.9114  \n",
       "22CO006     8.1447   8.7582  10.1497  10.4062  10.2491  10.3902  \n",
       "24CO005     7.8392   8.1849   9.6917   8.3264   7.8704  10.3740  \n",
       "27CO004     8.6110   9.1624   9.3106  11.0532   9.6777   9.6900  \n",
       "\n",
       "[106 rows x 13482 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RENAME the path to analysis_data!\n",
    "protein_data.to_csv(\"/Users/sabrinazhong/Desktop/USC/Spring_2022/QBIO490/qbio_data_analysis_sabrinaz/analysis_data/cptac/cptac_protein.csv\")\n",
    "protein_data_read_example = pd.read_csv(\"/Users/sabrinazhong/Desktop/USC/Spring_2022/QBIO490/qbio_data_analysis_sabrinaz/analysis_data/cptac/cptac_protein.csv\", index_col=0)\n",
    "protein_data_read_example\n",
    "\n",
    "# Repeat with clinical and RNA\n",
    "rna_data.to_csv(\"/Users/sabrinazhong/Desktop/USC/Spring_2022/QBIO490/qbio_data_analysis_sabrinaz/analysis_data/cptac/cptac_rna.csv\")\n",
    "rna_data_read_example = pd.read_csv(\"/Users/sabrinazhong/Desktop/USC/Spring_2022/QBIO490/qbio_data_analysis_sabrinaz/analysis_data/cptac/cptac_rna.csv\", index_col=0)\n",
    "rna_data_read_example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63c929b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
